### PR TITLE
add BigFloat negation and absolute value

### LINKED ISSUE #190: operate! for abs

### LINKED ISSUE #191: operate! for unary minus

### PR DESCRIPTION
**Pull request by nsajko**
Adapted from -(x::BigFloat) in Base.MPFR in mpfr.jl

Fixes #190
Fixes #191

### CODE REVIEW

File: src/implementations/BigFloat.jl [hunk 1]
```diff
@@ -78,6 +78,30 @@ function operate_to!(output::BigFloat, ::typeof(-), a::BigFloat, b::BigFloat)
...
 
+function operate!(::typeof(-), x::BigFloat)
+    ccall(
+        (:mpfr_neg, :libmpfr),
+        Int32,
+        (Ref{BigFloat}, Ref{BigFloat}, Base.MPFR.MPFRRoundingMode),
+        x,
+        x,
+        Base.MPFR.ROUNDING_MODE[])
+    x
```
Comment (blegat) [inline src/implementations/BigFloat.jl:89]:
```suggestion
    return x
```

File: src/implementations/BigFloat.jl [hunk 2]
```diff
@@ -78,6 +78,30 @@ function operate_to!(output::BigFloat, ::typeof(-), a::BigFloat, b::BigFloat)
...
+
+function operate!(::typeof(Base.abs), x::BigFloat)
+    ccall(
+        (:mpfr_abs, :libmpfr),
+        Int32,
+        (Ref{BigFloat}, Ref{BigFloat}, Base.MPFR.MPFRRoundingMode),
+        x,
+        x,
+        Base.MPFR.ROUNDING_MODE[])
+    x
```
Comment (blegat) [inline src/implementations/BigFloat.jl:102]:
```suggestion
    return x
```

**Comment by blegat** [comment]
Thanks, could you add tests ?

**Comment by nsajko** [comment]
My first Julia tests!

**Comment by nsajko** [comment]
Locally I ran the MA test suite with Julia nightly, and some tests were broken. I *guess* they're not related to this change:
```
Test Summary:       | Pass  Broken  Total   Time
@rewrite with Int64 | 2662       4   2666  34.5s
Test Summary:         | Pass  Broken  Total   Time
@rewrite with Float64 | 2662       4   2666  41.3s
Test Summary:        | Pass  Broken  Total   Time
@rewrite with BigInt | 2662       4   2666  47.9s
Test Summary:          | Pass  Broken  Total   Time
@rewrite with BigFloat | 2662       4   2666  52.7s
Test Summary:                 | Pass  Broken  Total   Time
@rewrite with Rational{Int64} | 2649       4   2653  54.9s
Test Summary:                  | Pass  Broken  Total   Time
@rewrite with Rational{BigInt} | 2649       4   2653  54.6s
Test Summary:             | Pass  Broken  Total   Time
@rewrite with DummyBigInt | 2644       4   2648  53.6s
```

```
Test Summary:                   | Pass  Broken  Total  Time
test_rewrite_nonconcrete_vector |    2       1      3  0.4s
```

**Comment by blegat** [comment]
Broken tests are fine, it's failing are errors that are worrying

**Comment by blegat** [comment]
See format-check failure

File: src/implementations/BigFloat.jl [hunk 3]
```diff
@@ -78,6 +78,30 @@ function operate_to!(output::BigFloat, ::typeof(-), a::BigFloat, b::BigFloat)
...
 end
 
+function operate!(::typeof(-), x::BigFloat)
+    ccall(
+        (:mpfr_neg, :libmpfr),
+        Int32,
+        (Ref{BigFloat}, Ref{BigFloat}, Base.MPFR.MPFRRoundingMode),
+        x,
+        x,
+        Base.MPFR.ROUNDING_MODE[])
```
Comment (odow) [inline src/implementations/BigFloat.jl:88]:
```suggestion
        Base.MPFR.ROUNDING_MODE[],
    )
```

File: src/implementations/BigFloat.jl [hunk 4]
```diff
@@ -78,6 +78,30 @@ function operate_to!(output::BigFloat, ::typeof(-), a::BigFloat, b::BigFloat)
...
+# Base.abs
+
+function operate!(::typeof(Base.abs), x::BigFloat)
+    ccall(
+        (:mpfr_abs, :libmpfr),
+        Int32,
+        (Ref{BigFloat}, Ref{BigFloat}, Base.MPFR.MPFRRoundingMode),
+        x,
+        x,
+        Base.MPFR.ROUNDING_MODE[])
```
Comment (odow) [inline src/implementations/BigFloat.jl:101]:
```suggestion
        Base.MPFR.ROUNDING_MODE[],
    )
```

File: test/bigfloat_neg_abs.jl [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2023 MutableArithmetics.jl contributors
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v.2.0. If a copy of the MPL was not distributed with this file, You can obtain
+# one at http://mozilla.org/MPL/2.0/.
+
+function neg_abs_test(f::F, x::BigFloat) where {F <: Function}
+    backup = x + 0
+    output = MA.operate!(f, x)
+    @test f(backup) == output == x
```
Comment (odow) [inline test/bigfloat_neg_abs.jl:10]:
```suggestion
    @test f(backup) == output == x
    return
```

File: test/bigfloat_neg_abs.jl [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2023 MutableArithmetics.jl contributors
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v.2.0. If a copy of the MPL was not distributed with this file, You can obtain
+# one at http://mozilla.org/MPL/2.0/.
+
+function neg_abs_test(f::F, x::BigFloat) where {F <: Function}
```
Comment (odow) [inline test/bigfloat_neg_abs.jl:7]:
```suggestion
function neg_abs_test(f::F, x::BigFloat) where {F<:Function}
```

File: test/bigfloat_neg_abs.jl [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    backup = x + 0
+    output = MA.operate!(f, x)
+    @test f(backup) == output == x
+end
+
+@testset "$input" for input in map(BigFloat, (-1.3, -1.0, -0.7, -0.0, 0.0, 0.3, 1.0, 1.7))
```
Comment (odow) [inline test/bigfloat_neg_abs.jl:13]:
```suggestion
@testset "$input" for input in map(
    BigFloat,
    (-1.3, -1.0, -0.7, -0.0, 0.0, 0.3, 1.0, 1.7),
)
```

**Review by blegat — APPROVED** [review]

**Pull request merged**